### PR TITLE
Fix causal mask in reference implementation of Attention

### PR DESCRIPTION
**Pull request by xadupre**
### Description
Fix causal mask for attention.

Another unrelated change was needed to achieve that goal: when the automated backend test produces the expanded version of a test (meaning the operator can be replaced by a function), it was done without considering the opset. This could lead to a wrong model. Now, the expanded test takes the function definition coming from the same opset than the one used in the test.

### Motivation and Context
Past should not be impacted by the causal mask.

### CODE REVIEW

File: docs/Changelog-ml.md [hunk 1]
```diff
@@ -609,7 +609,7 @@ This version of the operator has been available since version 1 of the 'ai.onnx.
 <dt><tt>T1</tt> : tensor(float), tensor(double), tensor(int64), tensor(int32)</dt>
 <dd>The input must be a tensor of a numeric type, either [C] or [N,C].</dd>
 <dt><tt>T2</tt> : tensor(string), tensor(int64)</dt>
-<dd>The output type will be a tensor of strings or integers, depending on which of the classlabels_* attributes is used. Its size will match the batch size of the input.</dd>
+<dd>The output type will be a tensor of strings or integers, depending on which of the classlabels_* attributes is used. Its size will match the bactch size of the input.</dd>
```
Comment (justinchuby) [inline docs/Changelog-ml.md]: Spell -> Looks like a merge conflict because this was recently fixed

**Comment by titaiwangms** [comment]
I wonder if we can make it in 1.19.1.

**Comment by titaiwangms** [comment]
Could you provide more information in the description. The idea that the prior implementation forgets to add offsets into masking would be enough.

File: docs/Changelog-ml.md (hunk 1 above)
Comment (xadupre) [reply docs/Changelog-ml.md]: it seems to have disappeared

File: onnx/reference/ops/op_attention.py [hunk 2]
```diff
@@ -24,6 +24,19 @@ def _softcap(X, softcap):
     return X
 
 
+def apply_causal(mask, inplace=False):
```
Comment (titaiwangms) [inline onnx/reference/ops/op_attention.py]:
```suggestion
def _apply_causal(mask, inplace=False):
```

For consistency

File: onnx/reference/ops/op_attention.py [hunk 3]
```diff
@@ -24,6 +24,19 @@ def _softcap(X, softcap):
...
+    q_sequence_length, total_sequence_length = mask.shape[-2:]
```
Comment (titaiwangms) [inline onnx/reference/ops/op_attention.py:33]: Please add a comment/function description explaining the process of this function.

File: onnx/reference/ops/op_attention.py [hunk 4]
```diff
@@ -24,6 +24,19 @@ def _softcap(X, softcap):
...
+    past_sequence_length = total_sequence_length - q_sequence_length
+    triu = np.triu(
+        np.ones((q_sequence_length, q_sequence_length), dtype=mask.dtype), k=1
+    )
+    triu[triu == 1] = -np.inf
+    if not inplace:
```
Comment (titaiwangms) [inline onnx/reference/ops/op_attention.py]: Is this really necessary for ref implementation?

Comment (titaiwangms) [reply onnx/reference/ops/op_attention.py]: In general, not inplace would be safer?

File: onnx/reference/ops/op_attention.py [hunk 5]
```diff
@@ -24,6 +24,19 @@ def _softcap(X, softcap):
     return X
 
 
+def apply_causal(mask, inplace=False):
+    q_sequence_length, total_sequence_length = mask.shape[-2:]
+    past_sequence_length = total_sequence_length - q_sequence_length
+    triu = np.triu(
+        np.ones((q_sequence_length, q_sequence_length), dtype=mask.dtype), k=1
+    )
+    triu[triu == 1] = -np.inf
```
Comment (titaiwangms) [inline onnx/reference/ops/op_attention.py:41]: When mask is bool array, you are trying to write float into it.

File: onnx/reference/ops/op_attention.py (hunk 2 above)
Comment (xadupre) [reply onnx/reference/ops/op_attention.py]: done

File: onnx/reference/ops/op_attention.py (hunk 5 above)
Comment (xadupre) [reply onnx/reference/ops/op_attention.py:41]: fixed

File: onnx/reference/ops/op_attention.py (hunk 3 above)
Comment (xadupre) [reply onnx/reference/ops/op_attention.py:33]: done

File: onnx/reference/ops/op_attention.py (hunk 4 above)
Comment (xadupre) [reply onnx/reference/ops/op_attention.py]: simplified

File: onnx/backend/test/case/node/__init__.py [hunk 6]
```diff
@@ -128,11 +129,25 @@ def rename_helper(internal_name: str) -> Any:
 
 
 def function_testcase_helper(
-    node: NodeProto, input_types: list[TypeProto], name: str
+    node: NodeProto,
+    input_types: list[TypeProto],
+    name: str,
+    opset_imports: Sequence[OperatorSetIdProto] | None = None,
```
Comment (titaiwangms) [inline onnx/backend/test/case/node/__init__.py:135]: I would mention this improvement in PR description.

File: onnx/defs/nn/utils.h [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+#include "onnx/defs/function.h"
+#include "onnx/defs/schema.h"
+
+namespace ONNX_NAMESPACE {
+namespace defs {
+namespace nn {
+namespace utils {
+
+/** Implements shape and type propagation for Attention (23-). */
+void AttentionPropagateElemTypeFromInputToOutput(InferenceContext& ctx);
```
Comment (titaiwangms) [inline onnx/defs/nn/utils.h:17]: Is there any change or you just move it here?

File: onnx/defs/nn/utils.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (titaiwangms) [inline onnx/defs/nn/utils.h:1]: utils.h and utils.cc is too general. Do you see them as a storage to save whatever functions here? Otherwise, we can make it more specific to attention? (Maybe attention_utils.h or attention_utils.cc)

File: onnx/defs/nn/utils.cc [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+      .Add("AttnBiasShape = Concat <axis = 0> (QSeqLen, NewKVSeqLen)");
+  float neg_inf = -std::numeric_limits<float>::infinity();
+  builder.Const1D("FloatNegInf", neg_inf);
+  builder.Const1D("ScalarZero", 0.f);
+
+  // If attn_mask is provided
+  if (ctx.hasInput(3)) {
+    auto* up = ctx.getInputType(3);
+    if ((up == nullptr) || (!up->has_tensor_type()))
+      return false;
```
Comment (titaiwangms) [inline onnx/defs/nn/utils.cc:173]: I see this is the only place we return false. What is the case we get None as attention_mask? And why do we fail in this case? Can't we treat it as not provided? I am confused here.

File: onnx/defs/nn/defs.cc [hunk 10]
```diff
@@ -3692,19 +3541,20 @@ ONNX_OPERATOR_SET_SCHEMA(
...
             // For 3D inputs: First reshape to [batch_size, seq_length, num_heads, head_size]
             // then transpose to [batch_size, num_heads, seq_length, head_size]
             builder
-                .Add("BatchSize = Shape <start = 0, end = 1> (Q)") // batch size
                 .Const1D("QNumHeadsAttr", q_num_heads) // q_num_heads from attrs
                 .Const1D("KVNumHeadsAttr", kv_num_heads) // kv_num_heads from attrs
-                .Add("QSeqLen = Shape <start = -2, end = -1> (Q)") // q_sequence_length
-                .Add("KVSeqLen = Shape <start = -2, end = -1> (K)") // kv_sequence_length
+                .Add("QSeqLen0 = Shape <start = -2, end = -1> (Q)") // q_sequence_length
+                .Add("KVSeqLen0 = Shape <start = -2, end = -1> (K)") // kv_sequence_length
```
Comment (titaiwangms) [inline onnx/defs/nn/defs.cc]: 0 seems to be a redundant change

**Review by titaiwangms — APPROVED** [review]
minor comments

File: onnx/defs/nn/utils.h (hunk 8 above)
Comment (justinchuby) [reply onnx/defs/nn/utils.h:1]: I think this has been a pattern in other directories, but I do agree with the point of having dedicated and focused files.

File: onnx/defs/nn/utils.h (hunk 7 above)
Comment (xadupre) [reply onnx/defs/nn/utils.h:17]: Just moved. It is common to old and new operator. I was hoping to do the same with the inline function but it looked like more complex to do.

File: onnx/defs/nn/utils.h (hunk 8 above)
Comment (xadupre) [reply onnx/defs/nn/utils.h:1]: It is the same pattern as the existing one. I suggest keeping it.

File: onnx/defs/nn/utils.cc (hunk 9 above)
Comment (xadupre) [reply onnx/defs/nn/utils.cc:173]: I did not change the existing code. Looking at the code, the mask can be boolean or float. If not known, this is no backup plan. This happens if shape inference fails at finding the type of the mask.

File: onnx/defs/nn/defs.cc (hunk 10 above)
Comment (xadupre) [reply onnx/defs/nn/defs.cc]: fixed

File: onnx/backend/test/case/node/__init__.py (hunk 6 above)
Comment (xadupre) [reply onnx/backend/test/case/node/__init__.py:135]: done

File: onnx/backend/test/case/node/__init__.py [hunk 11]
```diff
@@ -128,11 +129,25 @@ def rename_helper(internal_name: str) -> Any:
...
-    schema = onnx.defs.get_schema(test_op, domain=node.domain)
+    if opset_imports is None:
+        # No opset in the model. We take the most recent definition.
+        schema = onnx.defs.get_schema(test_op, domain=node.domain)
+    else:
+        # We take the function coming defined in the specific version mentioned
+        # in the model.
+        if len(opset_imports) != 1:
+            raise ValueError(
+                f"Only one domain is allowed but {len(opset_imports)} found."
```
Comment (justinchuby) [inline onnx/backend/test/case/node/__init__.py:147]: I think this will create issues with non-default opset ops. But we can think about this later.
